### PR TITLE
Add table oci_core_network_load_balancer. Closses #181

### PR DESCRIPTION
**Pull request by rajeshbal65**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
Test names: [ 'tests/oci_core_network_load_balancer' ]
No env file present for the current environment:  staging
 Falling back to .env config
No env file present for the current environment:  staging

SETUP: tests/oci_core_network_load_balancer []

PRETEST: tests/oci_core_network_load_balancer

TEST: tests/oci_core_network_load_balancer
Running terraform
oci_core_vcn.named_test_resource: Creating...
oci_core_vcn.named_test_resource: Creation complete after 2s [id=ocid1.vcn.oc1.ap-mumbai-1.amaaaaaa6igdexaasmvfaxyqxyqmfjozfxc723fb3dlzcpsvhkirh5qq4c5a]
oci_core_subnet.named_test_resource: Creating...
oci_core_subnet.named_test_resource: Creation complete after 2s [id=ocid1.subnet.oc1.ap-mumbai-1.aaaaaaaajqz4efnajrsxmsp2dk2k5nvxqawcpwjfu6jf2apckcrr2kjggzra]
oci_network_load_balancer_network_load_balancer.named_test_resource: Creating...
oci_network_load_balancer_network_load_balancer.named_test_resource: Still creating... [10s elapsed]
oci_network_load_balancer_network_load_balancer.named_test_resource: Still creating... [20s elapsed]
oci_network_load_balancer_network_load_balancer.named_test_resource: Creation complete after 21s [id=ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaaj4rlu4vdnou2a55vr62ov4lk7miy6rcw2a7knd43o5la]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.

Outputs:

display_name = "steampipetest4740"
freeform_tags = tomap({
  "Name" = "steampipetest4740"
})
is_preserve_source_destination = false
is_private = true
region = "ap-mumbai-1"
resource_id = "ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaaj4rlu4vdnou2a55vr62ov4lk7miy6rcw2a7knd43o5la"
resource_name = "steampipetest4740"
state = "ACTIVE"
subnet_id = "ocid1.subnet.oc1.ap-mumbai-1.aaaaaaaajqz4efnajrsxmsp2dk2k5nvxqawcpwjfu6jf2apckcrr2kjggzra"
tenancy_ocid = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"

Running SQL query: test-get-query.sql

[
  {
    "display_name": "steampipetest4740",
    "freeform_tags": {
      "Name": "steampipetest4740"
    },
    "id": "ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaaj4rlu4vdnou2a55vr62ov4lk7miy6rcw2a7knd43o5la",
    "is_private": true,
    "lifecycle_state": "ACTIVE",
    "subnet_id": "ocid1.subnet.oc1.ap-mumbai-1.aaaaaaaajqz4efnajrsxmsp2dk2k5nvxqawcpwjfu6jf2apckcrr2kjggzra"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql

[
  {
    "display_name": "steampipetest4740",
    "id": "ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaaj4rlu4vdnou2a55vr62ov4lk7miy6rcw2a7knd43o5la",
    "lifecycle_state": "ACTIVE",
    "subnet_id": "ocid1.subnet.oc1.ap-mumbai-1.aaaaaaaajqz4efnajrsxmsp2dk2k5nvxqawcpwjfu6jf2apckcrr2kjggzra"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql

null
✔ PASSED

Running SQL query: test-turbot-query.sql

[
  {
    "region": "ap-mumbai-1",
    "tenant_id": "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq",
    "title": "steampipetest4740"
  }
]
✔ PASSED

POSTTEST: tests/oci_core_network_load_balancer

TEARDOWN: tests/oci_core_network_load_balancer

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  display_name,
  id,
  subnet_id,
  lifecycle_state as state,
  health_status,
  time_created
from
  oci_core_network_load_balancer;

+--------------------+--------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------+--------+---------------+---------------------+
| display_name       | id                                                                                                     | subnet_id                                                                                 | state  | health_status | time_created        |
+--------------------+--------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------+--------+---------------+---------------------+
| nlb_2021-0719-1022 | ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaamgab5q5tvvgslnyzqwukcdcuwnomdi7aex47asazxtca | ocid1.subnet.oc1.ap-mumbai-1.aaaaaaaar45zy3bfclbig7m6xrfrrlwwgaojokqgq5htc2ezog26i7p2ugtq | ACTIVE | OK            | 2021-07-19 04:53:37 |
+--------------------+--------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------+--------+---------------+---------------------+
```

### List NLBs assigns with public IP address

```sql
select
  display_name,
  id,
  is_private
from
  oci_core_network_load_balancer
where
  not is_private;

+--------------------+--------------------------------------------------------------------------------------------------------+------------+
| display_name       | id                                                                                                     | is_private |
+--------------------+--------------------------------------------------------------------------------------------------------+------------+
| nlb_2021-0719-1022 | ocid1.networkloadbalancer.oc1.ap-mumbai-1.amaaaaaa6igdexaamgab5q5tvvgslnyzqwukcdcuwnomdi7aex47asazxtca | false      |
+--------------------+--------------------------------------------------------------------------------------------------------+------------+
```

### List unhealthy NLBs

```sql
select
  display_name,
  id,
  network_load_balancer_health -> 'status' as health_status
from
  oci_core_network_load_balancer
where
  network_load_balancer_health ->> 'status' = 'CRITICAL';

+--------------+----+---------------+
| display_name | id | health_status |
+--------------+----+---------------+
+--------------+----+---------------+
```
```
</details>

### CODE REVIEW

File: docs/tables/oci_core_network_load_balancer.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Table: oci_core_network_load_balancer
+
+The Network Load Balancing (NLB) feature distributes traffic across several servers by using the TCP/IP networking protocol.
```
Comment (bigdatasourav) [inline docs/tables/oci_core_network_load_balancer.md:3]:
```suggestion
Network Load Balancers can provide automated traffic distribution from one entry point to multiple servers in a backend set. It ensure that your services remain available by directing traffic only to healthy servers. Network Load Balancer provides the benefits of flow high availability, source and destination IP addresses, and port preservation. It is designed to handle volatile traffic patterns and millions of flows, offering high throughput while maintaining ultra low latency. It is the ideal load balancing solution for latency sensitive workloads. It is also optimized for long-running connections in the order of days or months.
```

File: docs/tables/oci_core_network_load_balancer.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  id,
+  subnet_id,
+  lifecycle_state as state,
+  health_status,
+  time_created
+from
+  oci_core_network_load_balancer;
+```
+
+
```
Comment (bigdatasourav) [inline docs/tables/oci_core_network_load_balancer.md:21]:
```suggestion
```

File: docs/tables/oci_core_network_load_balancer.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  display_name,
+  id,
+  is_private
+from
+  oci_core_network_load_balancer
+where
+  not is_private;
+```
+
+
```
Comment (bigdatasourav) [inline docs/tables/oci_core_network_load_balancer.md:35]:
```suggestion
```

File: docs/tables/oci_core_network_load_balancer.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+### List unhealthy NLBs
```
Comment (bigdatasourav) [inline docs/tables/oci_core_network_load_balancer.md:36]:
```suggestion
### List critical NLBs
```

File: docs/tables/oci_core_network_load_balancer.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+```sql
+select
+  display_name,
+  id,
+  health_status
+from
+  oci_core_network_load_balancer
+where
+  health_status <> 'OK';
```
Comment (bigdatasourav) [inline docs/tables/oci_core_network_load_balancer.md:46]:
```suggestion
  health_status = 'CRITICAL';
```

File: oci-test/tests/oci_core_network_load_balancer/test-turbot-expected.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "region": "ap-mumbai-1",
```
Comment (bigdatasourav) [inline oci-test/tests/oci_core_network_load_balancer/test-turbot-expected.json:3]: Don't hard code region. use output variable.

File: oci-test/tests/oci_core_network_load_balancer/variables.tf [hunk 7]
```diff
@@ -0,0 +1,86 @@
+variable "resource_name" {
+  type        = string
+  default     = "steampipetest20200125"
+  description = "Name of the resource used throughout the test."
+}
+
+variable "tenancy_ocid" {
+  type        = string
+  default     = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"
```
Comment (bigdatasourav) [inline oci-test/tests/oci_core_network_load_balancer/variables.tf:9]:
```suggestion
  default     = ""
```

File: oci/table_oci_core_network_load_balancer.go [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromCamel(),
+			},
+			{
+				Name:        "lifecycle_state",
+				Description: "The current state of the network load balancer.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "health_status",
```
Comment (bigdatasourav) [inline oci/table_oci_core_network_load_balancer.go:47]: It should be a json column, also need to update the query.

File: oci/table_oci_core_network_load_balancer.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+			},
+			{
+				Name:        "time_updated",
+				Description: "The date and time the network load balancer was created.",
+				Type:        proto.ColumnType_TIMESTAMP,
+				Transform:   transform.FromField("TimeUpdated.Time"),
+			},
+			{
+				Name:        "subnet_id",
+				Description: "The subnet in which the network load balancer is spawned OCIDs",
```
Comment (bigdatasourav) [inline oci/table_oci_core_network_load_balancer.go:72]:
```suggestion
				Description: "The subnet in which the network load balancer is spawned OCIDs.",
```

File: oci/table_oci_core_network_load_balancer.go [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+func getCoreNetworkLoadBalancerHealth(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getCoreNetworkLoadBalancerHealth")
+	logger := plugin.Logger(ctx)
+	region := plugin.GetMatrixItem(ctx)[matrixKeyRegion].(string)
+	compartment := plugin.GetMatrixItem(ctx)[matrixKeyCompartment].(string)
+	logger.Debug("getCoreNetworkLoadBalancerHealth", "Compartment", compartment, "OCI_REGION", region)
+
+	// Restrict the api call to only root compartment/ per region
+	if !strings.HasPrefix(compartment, "ocid1.tenancy.oc1") {
```
Comment (bigdatasourav) [inline oci/table_oci_core_network_load_balancer.go:243]: Will I get health details for NLBs which are not created in root?

**Review by bigdatasourav — CHANGES_REQUESTED** [review]
@rajeshbal65 plz check the requested changes

File: oci/table_oci_core_network_load_balancer.go [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromCamel(),
+			},
+			{
+				Name:        "lifecycle_state",
+				Description: "The current state of the network load balancer.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "network_load_balancer_health",
```
Comment (bigdatasourav) [inline oci/table_oci_core_network_load_balancer.go:47]: it should be after listeners, also sort the columns.

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Pull request merged**